core/trie/snap_support.go

```diff
@@ -0,0 +1,168 @@
+package trie
+
+import (
+	"github.com/NethermindEth/juno/core/felt"
+)
+
+func (t *Trie) IterateAndGenerateProof(startValue *felt.Felt, consumer func(key, value *felt.Felt) (bool, error),
+) ([]ProofNode, bool, error) {
+	var lastKey *felt.Felt
+
+	finished, err := t.Iterate(startValue, func(key, value *felt.Felt) (bool, error) {
+		lastKey = key
+
+		return consumer(key, value)
+	})
+	if err != nil {
+		return nil, false, err
+	}
+
+	proofset := map[felt.Felt]ProofNode{}
+
+	// If start value is null && finished, you dont need to provide any proof at all
+	if !finished || startValue != nil {
+		feltBts := startValue.Bytes()
+		startKey := NewKey(t.height, feltBts[:])
+		// Yes, the left proof is actually for the start query, not the actual leaf. Very confusing, yea I know. Need to
```

Comment:
What do you mean that the left proof is for the start query, and not the actual leaf? I thought the left proof would be for the first key/leaf in the range?

Reply:
Actually no. Its a very edge.. edge case. Can't remember at the top of my head why.